evms/evmone.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2023 Martin Holst Swende
+// This file is part of the goevmlab library.
+//
+// The library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the goevmlab library. If not, see <http://www.gnu.org/licenses/>.
+
+package evms
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/eth/tracers/logger"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type EvmoneVM struct {
+	path string
+	name string
+}
+
+func NewEvmoneVM(path string, name string) *EvmoneVM {
+	return &EvmoneVM{
+		path: path,
+		name: name,
+	}
+}
+
+func (evm *EvmoneVM) Name() string {
+	return fmt.Sprintf("evmone-%s", evm.name)
+}
+
+func (evm *EvmoneVM) GetStateRoot(path string) (root, command string, err error) {
+	cmd := exec.Command(evm.path, "--ignore-state-root", "--ignore-logs", "--trace", path)
+	data, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", cmd.String(), err
+	}
+	root, err = evm.ParseStateRoot(data)
+	if err != nil {
+		log.Error("Failed to find stateroot", "vm", evm.Name(), "cmd", cmd.String())
+		return "", cmd.String(), err
+	}
+	return root, cmd.String(), nil
+}
+
+func (evm *EvmoneVM) ParseStateRoot(data []byte) (root string, err error) {
+	start := bytes.Index(data, []byte(`"stateRoot": "`))
+	end := start + 14 + 66
+	if start == -1 || end >= len(data) {
+		return "", fmt.Errorf("%v: no stateroot found", evm.Name())
+	}
+	return string(data[start+14 : end]), nil
+}
+
+func (evm *EvmoneVM) RunStateTest(path string, out io.Writer, speedTest bool) (string, error) {
+	var (
+		stderr io.ReadCloser
+		err    error
+		cmd    *exec.Cmd
+	)
+
+	cmd = exec.Command(evm.path, "--ignore-state-root", "--ignore-logs", "--trace", path)
```

Comment:
I think we can handle "null" state root on evmone side without a flag. But I wonder, is there a good reason to pass "null" hash as `"0000..."` string instead of just omitting the field in JSON?

Reply:
What happens if you don't `--ignore-state-root` ?

Reply:
All vms complain about that, it's fine (as long as it also outputs what it expects)

Reply:
Example of geth complaining - the complaint is on stdout: https://github.com/holiman/goevmlab/blob/master/evms/testdata/traces/00000006-naivefuzz-0.json.geth.stdout.txt  (and the traces are on stderr)